shortfin/python/shortfin_apps/llm/components/generate.py

```diff
@@ -166,6 +167,12 @@ async def run(self):
             else:
                 input_batch = self.tokenize()
             for index, input_tokens in enumerate(input_batch):
+                decode_config = copy(self.decode_config)
+                decode_config.update_from_sampling_params(
+                    self.gen_req.sampling_params
+                    if self.gen_req.is_single
+                    else self.gen_req.sampling_params[index]
+                )
```

Comment:
Instead of copying every time, you could possibly only copy if there is a change in the sampling params:

```suggestion
                sampling_params = (
                    self.gen_req.sampling_params if self.gen_req.is_single 
                    else self.gen_req.sampling_params[index]
                )
                decode_config = (
                    copy(self.decode_config) if sampling_params 
                    else self.decode_config
                )
                if sampling_params:
                    decode_config.update_from_sampling_params(sampling_params)
```